Upstream has accepted a revised version that was adapted to changes in
git master that aren't in the 0.51.x branch:
https://github.com/mesonbuild/meson/commit/e5a67b42481f55295edc82f398509d760f07
d4ce

From 9591d349c19f7d8b87cc761787f4e8ef2417b48b Mon Sep 17 00:00:00 2001
From: Alan Coopersmith <alan.coopersmith@oracle.com>
Date: Tue, 1 Aug 2017 03:58:18 -0700
Subject: [PATCH] On Solaris, try gar for GNU ar

Needed on Solaris, where native ar doesn't recognize all the flags
that meson uses, but GNU ar does.

Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
---
 mesonbuild/environment.py | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/mesonbuild/environment.py b/mesonbuild/environment.py
index dc8f099b..c8822bf7 100644
--- a/mesonbuild/environment.py
+++ b/mesonbuild/environment.py
@@ -1183,6 +1183,8 @@ class Environment:
                 linkers = ['xilib']
             else:
                 linkers = [self.default_static_linker]
+            if mesonlib.is_sunos():
+                linkers.append('gar')
         popen_exceptions = {}
         for linker in linkers:
             if not {'lib', 'lib.exe', 'llvm-lib', 'llvm-lib.exe', 'xilib', 'xilib.exe'}.isdisjoint(linker):
-- 
2.15.2

